### PR TITLE
MSFT:53847109 [.NET Core] DTS / UserPrincipal.GetAuthorizationGroups returns PrincipalOperationException:  Information about the domain could not be retrieved (1212).

### PR DESCRIPTION
**Pull request by kumarravik78c**
UserPrincipal.GetAuthorizationGroups function returns exception when kerberos FAST is enabled which another security functionality. When FAST is enabled, the dot-net code does an additional domain look up with domainname "NT Authority" which is invalid since the domainname should not have a space in it. We are adding a check if the SID is sentinel type which meand that the sid is of the following Type then we do not do another domain look up.

Sentinel SID: S-1-5-21-X-Y-Z-R means that the SID belongs to a domain (including the local account domain) unless X=Y=Z=0 in which case it’s a sentinel SID, a special type of pseudo-object that can’t be interpreted in isolation.

### CODE REVIEW

**Comment by kumarravik78c** [comment]
@dotnet-policy-service agree [company="Microsoft"]

**Comment by kumarravik78c** [comment]
@dotnet-policy-service agree company="Microsoft"

**Review by jay98014 — APPROVED** [review]

**Comment by buyaa-n** [comment]
@kumarravik78c is it possible to add some level of testing?

File: src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs [hunk 1]
```diff
@@ -811,5 +817,58 @@ internal static bool IsMachineDC(string computerName)
...
+                // We return false as soon as we know the first subauthority is not 0
+                //
+                if (currentSubauthority != 0)
+                {
+                    return false;
+                }
+            }
+
+            //
+            // This means X=Y-Z=0
```
Comment (buyaa-n) [inline src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs:869]: NIT: might be intended to say X=Y=Z=0?
```suggestion
            // This means X=Y=Z=0
```

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs [hunk 2]
```diff
@@ -2324,5 +2330,58 @@ internal static IntPtr ConvertByteArrayToIntPtr(byte[] bytes)
...
+                // We return false as soon as we know the first subauthority is not 0
+                //
+                if (currentSubauthority != 0)
+                {
+                    return false;
+                }
+            }
+
+            //
+            // This means X=Y-Z=0
```
Comment (buyaa-n) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs:2382]: NIT: Ditto
```suggestion
            // This means X=Y=Z=0
```

File: src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs (hunk 1 above)
Comment (kumarravik78c) [reply src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs]: fixed. Thanks for catching this.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs [hunk 3]
```diff
@@ -2266,6 +2266,12 @@ internal static unsafe SidType ClassifySID(IntPtr pSid)
                 return SidType.FakeObject;
             }
 
+            // Is the SID S - 1 - 5 - 0 - 0 - 0 - RID(sentinel SID) ?
```
[the remark this answers is not in the document]
Comment (kumarravik78c) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: fixed.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs (hunk 2 above)
Comment (kumarravik78c) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: fixed.

**Comment by kumarravik78c** [comment]
> @kumarravik78c is it possible to add some level of testing?

Unfortunately, there is no unit testing framework around this code. The effort of adding the testing framework is too high for this fix.

File: src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs [hunk 4]
```diff
@@ -811,5 +817,58 @@ internal static bool IsMachineDC(string computerName)
                     Interop.Dsrole.DsRoleFreeMemory(dsRoleInfoPtr);
             }
         }
+
+        //
+        // The sentinel SID were placed in the domain SID range S-1-5-21-X-Y-Z-R with R < 512 because the existing domain controllers would always filter those SIDs out at boundaries.
+        // That way, the sentinel SID which says the claims or compound data is safe to consume would be removed should the claims or compound PAC ever pass through a domain controller
+        // that did not know how to apply security checks. S-1-5-21-X-Y-Z-R means that the SID belongs to a domain(including the local account domain) unless X=Y=Z=0 in which
+        // case it�s a sentinel SID, a special type of pseudo-object that can�t be interpreted in isolation.
```
Comment (buyaa-n) [inline src/libraries/System.DirectoryServices.AccountManagement/src/System/DirectoryServices/AccountManagement/Utils.cs:825]: NIT: Non ascii quotes
```suggestion
        // case it's a sentinel SID, a special type of pseudo-object that can't be interpreted in isolation.
```

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs [hunk 5]
```diff
@@ -2324,5 +2330,58 @@ internal static IntPtr ConvertByteArrayToIntPtr(byte[] bytes)
             Debug.Assert(pBytes != IntPtr.Zero);
             return pBytes;
         }
+
+        //
+        // The sentinel SID were placed in the domain SID range S-1-5-21-X-Y-Z-R with R < 512 because the existing domain controllers would always filter those SIDs out at boundaries.
+        // That way, the sentinel SID which says the claims or compound data is safe to consume would be removed should the claims or compound PAC ever pass through a domain controller
+        // that did not know how to apply security checks. S-1-5-21-X-Y-Z-R means that the SID belongs to a domain(including the local account domain) unless X=Y=Z=0 in which
+        // case it�s a sentinel SID, a special type of pseudo-object that can�t be interpreted in isolation.
```
Comment (buyaa-n) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs:2338]:
```suggestion
        // case it's a sentinel SID, a special type of pseudo-object that can't be interpreted in isolation.
```

**Review by buyaa-n — APPROVED** [review]
LGTM, thank you!

**Comment by kumarravik78c** [comment]
@buyaa-n , can you or someone else with write access merge this PR? I do not have write access. I am trying to figure out how to get the write access for this repo.

**Pull request merged**